cfme/networks/cloud_tenant.py

```diff
@@ -0,0 +1,72 @@
+import attr
+
+from cfme.cloud.tenant import TenantDetailsView
+from cfme.common import Taggable
+from cfme.modeling.base import BaseCollection, BaseEntity, parent_of_type
+from cfme.networks import ValidateStatsMixin
+from cfme.utils import version
+from cfme.utils.appliance.implementations.ui import navigator, CFMENavigateStep, navigate_to
+
+
+@attr.s
+class CloudTenant(Taggable, BaseEntity, ValidateStatsMixin):
+    """Class representing cloud tenants"""
+    in_version = ('5.10', version.LATEST)
+    category = "networks"
+    string_name = 'CloudTenant'
+    quad_name = None
+    db_types = ["CloudTenant"]
+
+    name = attr.ib()
+    provider_obj = attr.ib(default=None)
+
+    @property
+    def provider(self):
+        from cfme.networks.provider import NetworkProvider
+        return parent_of_type(self, NetworkProvider)
+
+    @property
+    def num_cloud_subnets(self):
+        view = navigate_to(self, "Details")
+        return int(view.entities.relationships.get_text_of("Cloud Subnets"))
+
+    @property
+    def num_network_routers(self):
+        view = navigate_to(self, "Details")
+        return int(view.entities.relationships.get_text_of("Network Routers"))
+
+    @property
+    def num_security_groups(self):
+        view = navigate_to(self, "Details")
+        return int(view.entities.relationships.get_text_of("Security Groups"))
+
+    @property
+    def num_network_ports(self):
+        view = navigate_to(self, "Details")
+        return int(view.entities.relationships.get_text_of("Network Ports"))
+
+
+@attr.s
+class CloudTenantCollection(BaseCollection):
+    """Collection object for CloudTenant object"""
+    ENTITY = CloudTenant
+
+    def all(self):
+        view = navigate_to(self.filters.get('parent'), 'CloudTenants')
+        cloud_tenants = view.entities.get_all(surf_pages=True)
+        return [self.instantiate(name=s.name) for s in cloud_tenants]
+
+
+@navigator.register(CloudTenant, 'Details')
+class Details(CFMENavigateStep):
+    VIEW = TenantDetailsView
+
+    def prerequisite(self, *args, **kwargs):
+        """Navigate through provider if it exists else navigate through parent object"""
+        if self.obj.provider:
+            return navigate_to(self.obj.provider, 'CloudTenants')
+        else:
+            return navigate_to(self.obj.parent, 'CloudTenants')
```

Comment:
The parent could well be a CloudTenantCollection, that will not have a CloudTenants destination registered. This destination is only registered against NetworkProvider.

Again this comes back to duplication of Tenant / CloudTenant when they appear to be the same entity in MIQ.

The navigation here needs to be structured so that navigation to a Tenant uses a prerequisite that accounts for the appropriate layers of the model. The parent of the `Tenant` is a `TenantCollection`, but that `TenantCollection` could have a parent of: `Appliance`, `CloudProvider`, `NetworkProvider`.  This parent could be due to how the class was created (`appliance.collections.cloud_tenants` vs `provider.collections.tenants`) or due to a filter applied to the collection after it was created.

Reply:
I pushed new changes, navigation now first checks if we have a provider or filter on collection and navigate through that. If we don't have provider or filter we navigate through default TenantCollection object `('Compute', 'Clouds', 'Tenants')`

I would kindly ask you to review our code again @mshriver :)